### DISCUSSION TITLE
<Link /> "as" path decorator strips away search params passed to "href"

### DISCUSSION THREAD

**Discussion by bashmentarium** [vercel/next.js#61864] [Help]
### Summary

I am currently using the "app" folder inside of a new v14 Next js project and I realized that using the "as" path decorator erases any query params you want to pass into the URL object to the "href" jsx attribute of a <Link /> component.

![Screenshot 2024-02-09 at 19 18 25](https://github.com/vercel/next.js/assets/56018612/19885aaa-9f66-432b-ba4a-2444fec58664)

----------------------------------------------------------------------------

![Screenshot 2024-02-09 at 19 30 39](https://github.com/vercel/next.js/assets/56018612/bb6765c0-57c9-49d2-af06-5a36254faf6a)

### Additional information

In a typical list/details type of application, the whole point of the "as" path decorator was to hide long UUIDs from the URL and replace them with tiny slug-ified names of products or whatever entities you had on your platform.

### Example

**UPDATE**: I have just run a test inside a fresh next boilerplate, this time using the old _getServerSideProps_ API and everything works just as it used to (and the way I remember it). 

<img width="1339" alt="Screenshot 2024-02-10 at 15 57 52" src="https://github.com/vercel/next.js/assets/56018612/1b5fdde6-07da-4647-9f8e-3d6fc7210b0b">

Apparently, this decorator only strips the query params in the new "app" folder way-of-things. I'll go back

**Comment by DevDaniloFerrari** [comment]
With the new router, the absence of the `as` parameter in various instances limits this capability. In such situations, it seems necessary to create a custom context for continued data sharing. Although the previous approach was quite handy for overcoming challenges in data sharing, especially between pages or consumers of useRouter, we now find ourselves in a position where we must acknowledge this and implement a more robust global state manager for such scenarios. While it may not be the ideal solution, it appears to be the current necessity.

**Reply by bashmentarium** [reply]
@DevDaniloFerrari Please, don't get it twisted, but your reply looks like some chatgpt generated lyrical essay that is not actually addressing the issue I have mentioned. I was expecting some help in the form of "do this" or "don't do this", or something at least remotely related to the way the "as" path decorator works, but what you wrote resembles some literary commentary on the issue I have opened. I have updated my post, so the only way to not get the query params stripped is to go back to the "**pages**" folder and stay away from the "app" folder for now. Cheers!

**Reply by DevDaniloFerrari** [reply]
Actually i've just got inspired by this [answer](https://github.com/vercel/next.js/discussions/49813#discussioncomment-5906395).

**Comment by bashmentarium** [comment]
The "as" decorator erases the query params passed to the "href" prop for any dynamic routes. Looks like Next has changed the way "as" works once again!

**Reply by icyJoseph** [reply]
Yeah, the `as` prop in App Router is very iffy... Could you try something, pass a `legacyBehavior` prop to the link. I am about to setup a project to do a bit of debugging, but if you could go ahead, and do some testing, that'd be great.

In all fairness, it is not even listed in the App Router docs, which is one of those, slightly annoying changes from Pages Router to App Router, such as lack of Router events, but unlike Router events, I am not sure there's a clear cut explanation for `as` misadventures.

**Reply by bashmentarium** [reply]
@icyJoseph I have just tried to pass legacyBehavior as a prop, and still the query params get stripped away
